v2/api/keyvault/customizations/vault_extensions.go

```diff
@@ -0,0 +1,289 @@
+			obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not been updated to match
+	var _ conversion.Hub = kv
+
+	// If createMode is nil, nothing for us to do
+	// (This shouldn't be possible, but better to hedge against it)
+	if kv.Spec.Properties == nil || kv.Spec.Properties.CreateMode == nil {
+		return armObj, nil
+	}
+
+	// Get the owner of the KeyVault, we need this resource group to determine the subscription
+	owner, ownerErr := ex.getOwner(ctx, kv, resolver, log)
+	if ownerErr != nil {
+		return nil, errors.Wrapf(ownerErr, "unable to find owner of KeyVault %s", kv.Name)
+	}
+
+	// Parse the ID of the owner
+	// (Can't use the KeyVault as we do this before the KV exists)
+	id, err := genruntime.GetAndParseResourceID(owner)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get and parse resource ID from KeyVault owner")
+	}
+
+	vc, err := armkeyvault.NewVaultsClient(id.SubscriptionID, armClient.Creds(), armClient.ClientOptions())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create new VaultsClient")
+	}
+
+	createMode := *kv.Spec.Properties.CreateMode
+	if createMode == CreateMode_CreateOrRecover {
+		createMode, err = ex.handleCreateOrRecover(ctx, kv, vc, resolver, log)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error checking for existence of soft-deleted KeyVault")
+		}
+	}
+
+	if createMode == CreateMode_PurgeThenCreate {
+		err = ex.handlePurgeThenCreate(ctx, kv, vc, resolver, log)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error purging soft-deleted KeyVault")
+		}
+
+		createMode = CreateMode_Default
+	}
+
+	// Modify the payload as necessary
+	spec := armObj.Spec()
+	err = reflecthelpers.SetProperty(spec, "Properties.CreateMode", &createMode)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error setting CreateMode to %s", createMode)
+	}
+
+	return armObj, nil
+}
+
+func (ex *VaultExtension) handleCreateOrRecover(
+	ctx context.Context,
+	kv *keyvault.Vault,
+	vc *armkeyvault.VaultsClient,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) (string, error) {
+	exists, err := ex.checkForExistenceOfDeletedKeyVault(ctx, kv, resolver, vc, log)
+	if err != nil {
+		return "", errors.Wrapf(err, "error checking for existence of soft-deleted KeyVault %s", kv.Name)
+	}
+
+	result := CreateMode_Default
+	if exists {
+		result = CreateMode_Recover
+	}
+
+	log.Info(
+		"KeyVault reconciliation requested CreateOrRecover",
+		"KeyVault", kv.Name,
+		"softDeletedKeyvaultExists", exists,
+		"createMode", result)
+
+	return result, err
+}
+
+func (ex *VaultExtension) handlePurgeThenCreate(
+	ctx context.Context,
+	kv *keyvault.Vault,
+	vc *armkeyvault.VaultsClient,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) error {
+	// Find out whether a soft-deleted KeyVault with the same name exists
+	exists, err := ex.checkForExistenceOfDeletedKeyVault(ctx, kv, resolver, vc, log)
+	if err != nil {
```

Comment:
Since KV names should be globally unique - Not sure what happens in the case where the soft-deleted KV is in another subscription and it's just the name that is not available. 

Just double-checking if the user would get the `name not available` event/log anyway ?